Add Mbed TLS Support

Sean-Der opened the pull request:
This is still a WIP. If anyone wants to follow my progress/has feedback just want to make it public

paullouisageneau commented on CMakeLists.txt:
```diff
@@ -302,6 +308,15 @@ else()
 
+if (USE_GNUTLS OR USE_MBEDTLS)
+	if (NOT NO_WEBSOCKET)
+		# Needed for SHA1, it should be present as GnuTLS/MbedTLS cryptography backend
```
Unless I'm mistaken, MbedTLS does not rely on Nettle, so Nettle should not be linked against in that case. Instead, it looks like the SHA1 API is in `mbedtls/sha1.h`.

Sean-Der replied:
:+1: Fixed, removed the nettle usage for mbed tls

paullouisageneau commented on CMakeLists.txt:
```diff
@@ -17,7 +18,12 @@ option(WARNINGS_AS_ERRORS "Treat warnings as errors" OFF)
+endif()
+
+
+if(USE_GNUTLS OR USE_MBEDTLS)
```
Same here, Nettle should not be linked against with MbedTLS.

Sean-Der replied:
:+1: Fixed, removed the nettle usage for mbed tls

tytan652 commented on CMakeLists.txt:
```diff
@@ -317,11 +323,18 @@ if (USE_GNUTLS)
 		target_link_libraries(datachannel-static PRIVATE Nettle::Nettle)
 	endif()
+elseif(USE_MBEDTLS)
+	find_package(MbedTLS CONFIG REQUIRED)
```
```suggestion
	find_package(MbedTLS REQUIRED)
```
I don't think `CONFIG` is needed.

A finder should be added to the project since MbedTLS, since neither CMake or MbedTLS do and it might help you for checking for needed components (e.g. X509). 

Here [one](https://github.com/PatTheMav/obs-studio/blob/macos-build-update/cmake/finders/FindMbedTLS.cmake), you could use. It's the one plan to be used OBS Studio in the future, some OBS repo related comments could be removed.

tytan652 replied:
Side-note if there any issues with the finder don't hesitate to report them there: https://github.com/obsproject/obs-studio/pull/8247

Sean-Der replied:
@RytoEX Fixed! Thanks for calling that out.

tytan652 reviewed (commented):
I think this PR needs to be tested against MbedTLS 2 and 3.

Sean-Der commented on CMakeLists.txt:
```diff
@@ -317,11 +323,18 @@ if (USE_GNUTLS)
+	find_package(MbedTLS CONFIG REQUIRED)
+
+	target_compile_definitions(datachannel PRIVATE USE_MBEDTLS=1)
+	target_compile_definitions(datachannel-static PRIVATE USE_MBEDTLS=1)
```
Fixed!

Sean-Der replied:
Reverted. This broke the build

Sean-Der commented:
@tytan652 What version of mbedtls does OBS use today?

I am hesitant to support mbed tls because it has things like `DTLS 1.0`. I don't want to encourage/support insecure endpoints https://bugs.chromium.org/p/webrtc/issues/detail?id=10261

tytan652 commented:
But you can support only 3 if you want and use GNUTLS for Linux, we already do that for FFmpeg (one different TLS lib per OS).

Sean-Der commented:
Thanks @tytan652. Your choice @paullouisageneau it is your code base! 

I can add support for both (if you prefer), but personally I prefer less ifdef and lines of code and push people to update dependencies.

paullouisageneau commented:
@Sean-Der As I understand, there is no runtime setting in MbedTLS to force a minimal TLS version? That sounds like a weird design.

In that case, I think you may support only version 3, as said earlier people can link against GnuTLS (or OpenSSL) system libraries on Linux.

Sean-Der commented:
This compiles/all the WebRTC tests now pass! 

I need to implement the Websocket side now. Did this with `-DNO_WEBSOCKET`

paullouisageneau commented:
> This compiles/all the WebRTC tests now pass!
> 
> I need to implement the Websocket side now. Did this with `-DNO_WEBSOCKET`

Great news, congrats! Could you please add a Github worflow `build-mbedtls.yml` when you have finished with the TLS transport?

Sean-Der commented:
@paullouisageneau done!

I am gonna head to bed. In the morning will fix any test failures that have come up. When everything passes I think we are ready for a review!

tytan652 commented on .gitmodules:
```diff
@@ -17,3 +17,6 @@
 	path = deps/libsrtp
 	url = https://github.com/cisco/libsrtp.git
 	shallow = true
+[submodule "deps/mbedtls"]
+	path = deps/mbedtls
+	url = https://github.com/Mbed-TLS/mbedtls.git
```
Is Mbed TLS added as a submodule just because Ubuntu only provides the version 2 ?

For the macOS, you use homebrew like for OpenSSL.

Sean-Der replied:
Unfortunately I can't use Mbed TLS from homebrew or apt.

`MBEDTLS_SSL_DTLS_SRTP` isn't enabled by default. I opened a PR [here](https://github.com/Homebrew/homebrew-core/pull/125402) to enabled it for homebrew.

Sean-Der replied:
@paullouisageneau I agree, but I think this is the 'least' worst?

* Users that don't enable Mbed TLS aren't paying a cost (add_subdirectory doesn't happen)
* Keeps CI+Development simple. Having a pinned version of Mbed TLS and making development easy is nice

In the future if Homebrew and Ubuntu package Mbed TLS with DTLS-SRTP enabled we could drop it then?

paullouisageneau replied:
There is still a cost for users because Mbed TLS will be cloned by default, for instance when you run `git clone --recursive` or `git submodule --init`.  Since the Homebrew package now has DTLS-SRTP enabled (thank you for this), I think for now you could remove Ubuntu which doesn't feature MbedTLS 3 anyway and drop the submodule.

Sean-Der replied:
@paullouisageneau Done! Removed as submodule, and will install as system lib in github action.

tytan652 commented:
A CMake finder for Mbed TLS is still missing, CMake and Mbed TLS does not provide their own. You need to provide to libdatachannel this finder.

tytan652 commented on CMakeLists.txt:
```diff
@@ -335,6 +342,20 @@ if (USE_GNUTLS)
 	endif()
+elseif(USE_MBEDTLS)
+	if(USE_SYSTEM_MBEDTLS)
+		find_package(MbedTLS REQUIRED)
```
You must set a minimum version in the `find_package`. Replace the `3` by right minimum version if needed.
```suggestion
		find_package(MbedTLS 3 REQUIRED)
```

Sean-Der commented:
@paullouisageneau All the tests pass!

We have some duplication across GnuTLS/OpenSSL/MbedTLS. I would love to start another PR after this to clean that up a bit (if you want/agree)

paullouisageneau commented:
> @paullouisageneau All the tests pass!

Looking great, I'll do I thorough review when I get a moment.

> We have some duplication across GnuTLS/OpenSSL/MbedTLS. I would love to start another PR after this to clean that up a bit (if you want/agree)

It would be nice, thank you!

Sean-Der commented:
@paullouisageneau At a high level anything I need to improve?

Also feel free to cherry-pick/rewrite/split commits etc... as a maintainer it can be frustrating to get code that isn't consistent with the code base, so totally understand if you want to rewrite everything :)

DDRBoxman commented on cmake/Modules/FindMbedTLS.cmake:
```diff
@@ -0,0 +1,208 @@
+The following cache variables may also be set:
+
+``MbedTLS_<COMPONENT>_LIBRARY``
+  Path to the library component of FFmpeg.
```
FFmpeg -> mbedtls

Sean-Der replied:
Fixed! Thanks @DDRBoxman :)

Sean-Der commented:
@paullouisageneau rebased!

I don't see Mbed TLS tests here, but I opened up a PR against my fork and everything passed. Since I am not a owner of this repo I can't add new jobs?

tytan652 commented on CMakeLists.txt:
```diff
@@ -335,6 +342,20 @@ if (USE_GNUTLS)
+
+	target_compile_definitions(datachannel PRIVATE USE_MBEDTLS)
+	target_compile_definitions(datachannel-static PRIVATE USE_MBEDTLS)
+	target_link_libraries(datachannel PRIVATE MbedTLS::mbedtls)
+	target_link_libraries(datachannel-static PRIVATE MbedTLS::mbedtls)
```
The Finder provides `MbedTLS::MbedTLS` and not `MbedTLS::mbedtls`.

Sean-Der replied:
@tytan652 That breaks the build

```
CMake Error at CMakeLists.txt:357 (target_link_libraries):
  Target "datachannel" links to:

    MbedTLS::MbedTLS

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:358 (target_link_libraries):
  Target "datachannel-static" links to:

    MbedTLS::MbedTLS

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```

tytan652 replied:
Because the submodules and the Finder for system install set different targets. So you need to use `MbedTLS::mbedtls` if `USE_SYSTEM_MBEDTLS=OFF` and `MbedTLS::MbedTLS` if `USE_SYSTEM_MBEDTLS=ON`.

Sean-Der replied:
Fixed! Mbed TLS is not a submodule anymore, and use `MbedTLS::MbedTLS` properly.

paullouisageneau commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,308 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+		if (message->size() > size_t(mbedtls_ssl_get_max_out_record_payload(&mSsl)))
+			return false;
+
+		auto buff = &(message->data())[0];
+
+		ret = mbedtls_ssl_write(&mSsl, (const unsigned char *) buff, message->size());
```
```suggestion
		ret = mbedtls_ssl_write(&mSsl, reinterpret_cast<const unsigned char *>(message->data()), message->size());
```

paullouisageneau commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,308 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+					PLOG_DEBUG << "DTLS connection cleanly closed";
+					break;
+				}
+				switch (ret) {
+					case MBEDTLS_ERR_SSL_WANT_READ:
+					case MBEDTLS_ERR_SSL_WANT_WRITE:
+					case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+					case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS:
+					case MBEDTLS_ERR_SSL_CLIENT_RECONNECT:
+						return;
+				}
```
I would use a simple if here for readability. Also, do all of those return codes make sense here? For instance, shouldn't the connection be closed instead on client reconnect?

Sean-Der replied:
`MBEDTLS_SSL_RENEGOTIATION_DISABLED` is the default, so we don't need to handle it ourselves. Converted the other statements to an if.

paullouisageneau reviewed (commented):
I reviewed on left a bunch of comments, on the whole you did a great job on this, thank you for the contribution!

Sean-Der commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,308 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+	enqueueRecv();
+}
+
+void DtlsTransport::SetTimerCallback(void *ctx, uint32_t int_ms, uint32_t fin_ms) {
```
Moved!

Not a nitpick IMO :) you have to maintain this code, so happy to modify this PR in any way that makes it easier.

Sean-Der commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,308 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+                                         const unsigned char client_random[32],
+                                         const unsigned char server_random[32],
+                                         mbedtls_tls_prf_types tls_prf_type) {
+	(void) type;
```
Fixed! I didn't know you could just comment out the variable name, I have been doing it wrong :)

Sean-Der commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,308 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+	mbedtls_ctr_drbg_init(&mDrbg);
+	mbedtls_ssl_init(&mSsl);
+	mbedtls_ssl_config_init(&mConf);
+	mbedtls_ctr_drbg_set_prediction_resistance(&mDrbg, MBEDTLS_CTR_DRBG_PR_ON);
```
Added to certificate generation.

Sean-Der commented on src/impl/certificate.cpp:
```diff
@@ -149,7 +149,184 @@ string make_fingerprint(gnutls_x509_crt_t crt) {
+
+	mbedtls::check(mbedtls_sha256(crt->raw.p, crt->raw.len, (unsigned char *) buffer, 0));
+
+	auto size = mbedtls_md_get_size(mbedtls_md_info_from_type(MBEDTLS_MD_SHA256));
```
Switched to `size = 32`

I like defining the size up front better. When I read the code I don't have to assume/guess the return values

Sean-Der commented on src/impl/certificate.cpp:
```diff
@@ -149,7 +149,184 @@ string make_fingerprint(gnutls_x509_crt_t crt) {
+	    });
+
+  mbedtls_x509_crt_init(crt.get());
+	mbedtls::check(mbedtls_x509_crt_parse(crt.get(), (const unsigned char *) crt_pem.c_str(), crt_pem.length()));
```
Fixed! Created `mbedtls::new_x509_crt()` and `mbedtls::new_pk_context()`

Sean-Der commented on src/impl/certificate.cpp:
```diff
@@ -149,7 +149,184 @@ string make_fingerprint(gnutls_x509_crt_t crt) {
+	    });
+
+	mbedtls_x509_crt_init(crt.get());
+	mbedtls::check(mbedtls_x509_crt_parse_file(crt.get(), crt_pem_file.c_str()));
```
Fixed! Created `mbedtls::new_x509_crt()` and `mbedtls::new_pk_context()`

Sean-Der commented:
@paullouisageneau I believe I have addressed all the comments. Could I get another review please?

thank you!

pkviet commented:
Tried to build on visual studio 2022 without success.
I have include errors like:
```
libdatachannel\deps\libsrtp\crypto\include\aes_gcm.h(69,10): fatal  error C1083: Cannot open include file: 'mbedtls/aes.h': No such file or directory
libdatachannel\deps\libsrtp\crypto\cipher\aes_icm_mbedtls.c(48,10): fatal  error C1083: Cannot open include file: 'mbedtls/aes.h': No such file or directory
libdatachannel\deps\libsrtp\crypto\cipher\aes_gcm_mbedtls.c(49,10): fatal  error C1083: Cannot open include file: 'mbedtls/gcm.h': No such file or directory
libdatachannel\deps\libsrtp\crypto\hash\hmac_mbedtls.c(52,10): fatal  error C1083: Cannot open include file: 'mbedtls/md.h': No such file or directory
```

pkviet commented:
Hi @Sean-Der 
I first tested through Colin's pr on obs-deps.
It failed for me and someone else, the error requested a flag to enable c++20. Tried that to no avail. Didn't fix anything.
After that failure, I tried building directly your libdatachannel PR on visual studio without modifications and came up with the include errors in libsrtp in addition to c++20 errors. 
I've tried to find fixes but was unsuccessful.
So I've told Colin who'll try to debug tonight.

pkviet commented:
Oh great @Sean-Der !

Sean-Der commented:
ok pushed, @pkviet would you mind trying 10ad7beeff6218280f6c6521d40230379858c2e5

This is the change I made. The libsrtp CMakeLists was effected by our finder

```
diff --git a/cmake/Modules/FindMbedTLS.cmake b/cmake/Modules/FindMbedTLS.cmake
index 3b13876..a82ba04 100644
--- a/cmake/Modules/FindMbedTLS.cmake
+++ b/cmake/Modules/FindMbedTLS.cmake
@@ -193,9 +193,11 @@ if(MbedTLS_FOUND)

   if(MbedTLS_INCLUDES_X509 AND MbedTLS_INCLUDES_CRYPTO)
     set(MbedTLS_LIBRARIES ${MbedTLS_LIBRARY})
+    set(MBEDTLS_INCLUDE_DIRS ${MbedTLS_INCLUDE_DIR})
   else()
     set(MbedTLS_LIBRARIES ${MbedTLS_LIBRARY} ${MbedCrypto_LIBRARY} ${MbedX509_LIBRARY})
     set_property(TARGET MbedTLS::MbedTLS PROPERTY INTERFACE_LINK_LIBRARIES MbedTLS::MbedCrypto MbedTLS::MbedX509)
+    set(MBEDTLS_INCLUDE_DIRS ${MbedTLS_INCLUDE_DIR})
   endif()
 endif()
```

pkviet commented:
Will do, will report here and on obs discord.

pkviet commented:
so the include errors for libsrtp are gone but i still have a bunch of these:
`5>certificate.cpp
5>L:\visual-studio\obs-deps\windows\obs-deps-x64\include\psa\crypto_extra.h(1948,43): error C7555: use of designated initializers requires at least '/std:c++20' `
These includes come from mbedtls.

DDRBoxman commented:
Okay, so some updates about the build errors for OBS deps on windows.

Just for kicks I tried mbedtls 3.2.1, but it has compile issues in certificate.cpp which I assume is related to the recent change Sean made.
`\libdatachannel\src\impl\certificate.cpp(157,17): error C3861: 'mbedtls_sha256': identifier not found`
It does not show the designated initializers issue though.

Turns out that the designated initializers being an issue in C++ projects is a known issue in mbedtls that might get fixed next quarter https://github.com/Mbed-TLS/mbedtls/issues/7087

Sean-Der commented:
@DDRBoxman is that the only error it has with `mbedtls_sha256`? That feels really close?

Maybe we are passing the wrong types, are that isn't enabled?

tytan652 commented on cmake/Modules/FindMbedTLS.cmake:
```diff
@@ -0,0 +1,210 @@
+  unset(CMAKE_REQUIRED_INCLUDES)
+endif()
+
+if(CMAKE_HOST_SYSTEM_NAME MATCHES "Darwin|Windows")
+  set(MbedTLS_ERROR_REASON "Ensure that obs-deps is provided as part of CMAKE_PREFIX_PATH.")
+elseif(CMAKE_HOST_SYSTEM_NAME MATCHES "Linux|FreeBSD")
+  set(MbedTLS_ERROR_REASON "Ensure that MbedTLS is installed on the system.")
+endif()
```
```suggestion
```

tytan652 commented on cmake/Modules/FindMbedTLS.cmake:
```diff
@@ -0,0 +1,210 @@
+  find_package_handle_standard_args(
+    MbedTLS
+    REQUIRED_VARS MbedTLS_LIBRARY MbedCrypto_LIBRARY MbedX509_LIBRARY MbedTLS_INCLUDE_DIR
+    VERSION_VAR MbedTLS_VERSION REASON_FAILURE_MESSAGE "${MbedTLS_ERROR_REASON}")
```
```suggestion
    VERSION_VAR MbedTLS_VERSION)
```

tytan652 commented on cmake/Modules/FindMbedTLS.cmake:
```diff
@@ -0,0 +1,210 @@
+  mark_as_advanced(MbedTLS_LIBRARY MbedCrypto_LIBRARY MbedX509_LIBRARY MbedTLS_INCLUDE_DIR)
+  list(APPEND _COMPONENTS TLS Crypto X509)
+endif()
+unset(MbedTLS_ERROR_REASON)
```
```suggestion
```

tytan652 reviewed (changes requested):
`REASON_FAILURE_MESSAGE` is not needed.

paullouisageneau commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,303 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+					ret == MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS ||
+					ret == MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS)
+			{
+				auto timeout = milliseconds(this->mFin_ms);
```
Shouldn't the timeout be set to the remaining duration by accounting for time already elapsed since `mTimerSetAt`?

Sean-Der replied:
Yep you are right! Took into account the elapsed time

paullouisageneau commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,303 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+
+		ret = mbedtls_ssl_write(&mSsl, reinterpret_cast<const unsigned char *>(message->data()), message->size());
+	} while (ret == MBEDTLS_ERR_SSL_WANT_WRITE);
+	mbedtls::check(ret, "Failed writing to Mbed TLS Context");
```
```suggestion
	mbedtls::check(ret);
```
(for consistency with other implementations and TLS)

paullouisageneau commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,303 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+				return;
+			}
+
+			mbedtls::check(ret, "Failed Mbed DTLS Handshake");
```
```suggestion
			mbedtls::check(ret, "DTLS handshake failed");
```

paullouisageneau commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,303 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+				{
+					return;
+				}
+				mbedtls::check(ret, "Failed Mbed DTLS Read");
```
```suggestion
				mbedtls::check(ret);
```

paullouisageneau commented on src/impl/dtlstransport.hpp:
```diff
@@ -70,7 +70,35 @@ class DtlsTransport : public Transport, public std::enable_shared_from_this<Dtls
+	mbedtls_ssl_config mConf;
+	mbedtls_ssl_context mSsl;
+
+	uint32_t mFin_ms, mInt_ms;
```
I'd prefer `mFinMs` and `mIntMs` to prevent mixing camel-case and snake-case.

Sean-Der replied:
Fixed! Renamed

paullouisageneau commented on src/impl/tlstransport.cpp:
```diff
@@ -292,7 +292,223 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+		}
+
+		if (message) {
+			size_t available = message->size() - position;
+			ssize_t writeLen = std::min((unsigned long) len, available);
+			std::memcpy(buf, message->data() + position, writeLen);
+			position += len;
+			return len;
```
I think the wrong size if added and returned.
```suggestion
			size_t available = message->size() - position;
			size_t writeLen = std::min(len, available);
			std::memcpy(buf, message->data() + position, writeLen);
			position += writeLen;
			return writeLen;
```

Sean-Der commented on src/impl/certificate.cpp:
```diff
@@ -149,7 +148,149 @@ string make_fingerprint(gnutls_x509_crt_t crt) {
+		std::memset(certificateBuffer, 0, certificateBufferSize);
+
+		auto certificateLen = mbedtls_x509write_crt_der(&wcrt, certificateBuffer, certificateBufferSize, mbedtls_ctr_drbg_random, &drbg);
+		if (certificateLen <= 0){
```
Ran clang-format

Sean-Der commented:
Hey @paullouisageneau 

Addressed all the issues and rebased, mind giving another review? thanks!

jordo commented:
(Just a comment from the sidelines), but thanks for all the hard work on this everyone BTW... I'm really looking forward to trying this out in our games!

Sean-Der commented:
@paullouisageneau anything else I can do/change?

Also after this PR would you be interested in me adding sanitizers? Running TSAN locally I noticed some things around logging etc...

Sean-Der commented:
Hey @paullouisageneau 

Would it help to get another reviewer? Anyone else involved in libdatachannel I should ping?

thanks!

paullouisageneau commented on src/impl/tlstransport.cpp:
```diff
@@ -296,7 +296,224 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+		// Handle handshake if connecting
+		if (state() == State::Connecting) {
+			auto handshakeStatus = mbedtls_ssl_handshake(&mSsl);
+			switch (handshakeStatus) {
+			case 0:
+				break;
+			case MBEDTLS_ERR_SSL_WANT_WRITE:
+			case MBEDTLS_ERR_SSL_WANT_READ:
+			case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+			case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS: {
```
Could you please refactor this the same was as in `DtlsTransport`?

Sean-Der replied:
Fixed!

paullouisageneau commented on src/impl/tlstransport.cpp:
```diff
@@ -296,7 +296,224 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+					PLOG_DEBUG << "TLS connection cleanly closed";
+					break;
+				}
+				switch (ret) {
+				case MBEDTLS_ERR_SSL_WANT_READ:
+				case MBEDTLS_ERR_SSL_WANT_WRITE:
+				case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+				case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS:
+				case MBEDTLS_ERR_SSL_CLIENT_RECONNECT:
+					return;
```
Could you please refactor this the same was as in `DtlsTransport`?

paullouisageneau commented on src/impl/tlstransport.cpp:
```diff
@@ -296,7 +296,224 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+			case MBEDTLS_ERR_SSL_WANT_READ:
+			case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+			case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS: {
+				auto timeout = milliseconds(100);
+				ThreadPool::Instance().schedule(timeout, [weak_this = weak_from_this()]() {
+					if (auto locked = weak_this.lock())
+						locked->doRecv();
+				});
```
This should be removed as the connection is reliable and there are no retransmissions here.

Sean-Der replied:
I believe we still need this @paullouisageneau when running locally I would occasionally see handshake failures without it.  

```
   * \return         #MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS if a cryptographic
   *                 operation is in progress (see mbedtls_ecp_set_max_ops()) -
   *                 in this case you must call this function again to complete
   *                 the handshake when you're done attending other tasks.
```

paullouisageneau replied:
Calling again after an arbitrary timeout is kind of a hack and could indicate a return code is not handled as expected currently. Do you still see handshake failures after fixing `WriteCallback` not to incorrectly return `MBEDTLS_ERR_SSL_WANT_READ`? This could have caused the reception to wait for input and stall instead of reading available incoming data.

If I understand the doc correctly, the `X_IN_PROGRESS` functions require the user to call again as soon as possible, so there should be no wait in those cases (same for DTLS). Also, it sounds like `MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS` may only be returned if `mbedtls_ecp_set_max_ops()` was called to limit the number of operations before interrupting as the default is unlimited, so I guess it's never returned here.

Sean-Der replied:
I still failures after the `WriteCallback` changed

I updated the code to call again immediately on the suggest return codes (instead of timing out)

paullouisageneau commented on src/impl/tlstransport.cpp:
```diff
@@ -296,7 +296,224 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+int TlsTransport::WriteCallback(void *ctx, const unsigned char *buf, size_t len) {
+	auto *t = static_cast<TlsTransport *>(ctx);
+	auto *b = reinterpret_cast<const byte *>(buf);
+	if (!t->outgoing(make_message(b, b + len))) {
+		return MBEDTLS_ERR_SSL_WANT_READ;
+	}
```
```suggestion
	t->outgoing(make_message(b, b + len));
```

paullouisageneau commented on src/impl/tlstransport.cpp:
```diff
@@ -296,7 +296,224 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+	enqueueRecv();
+}
+
+bool TlsTransport::outgoing(message_ptr message) { return Transport::outgoing(std::move(message)); }
```
```suggestion
bool TlsTransport::outgoing(message_ptr message) {
	bool result = Transport::outgoing(std::move(message));
	mOutgoingResult = result;
	return result;
}
```

paullouisageneau commented on src/impl/tlstransport.cpp:
```diff
@@ -296,7 +296,224 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+	mbedtls::check(mbedtls_ssl_write(
+	    &mSsl, reinterpret_cast<const unsigned char *>(message->data()), int(message->size())));
+
+	return true;
```
```suggestion
	return mOutgoingResult;
```

paullouisageneau commented:
Here is my (hopefully last) review, I think after addressing the comments this should be good to go.

> Also after this PR would you be interested in me adding sanitizers? Running TSAN locally I noticed some things around logging etc...

Yes sure!

> Also after this is merged will it be tagged? I am happy to grab some other bugs/issues to warrant a new release. The OBS build system requires that it be tagged.

Sure, I'll create a new release with it. Are there other changes needed for OBS?

Sean-Der commented on src/impl/certificate.cpp:
```diff
@@ -149,7 +147,173 @@ string make_fingerprint(gnutls_x509_crt_t crt) {
+		}
+
+		auto now =
+		    std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now());
```
Removed cast

Sean-Der commented on src/impl/dtlstransport.hpp:
```diff
@@ -70,7 +70,33 @@ class DtlsTransport : public Transport, public std::enable_shared_from_this<Dtls
+	mbedtls_ssl_config mConf;
+	mbedtls_ssl_context mSsl;
+
+	uint32_t mFinMs, mIntMs;
```
done

Sean-Der commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,305 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+			if (ret == MBEDTLS_ERR_SSL_WANT_READ || ret == MBEDTLS_ERR_SSL_WANT_WRITE ||
+			    ret == MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS ||
+			    ret == MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS) {
+				auto elapsed = std::chrono::steady_clock::now() - mTimerSetAt;
```
Done! Condensed into one liner

Sean-Der commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,305 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+
+	if (dtlsTransport->mFinMs == 0) {
+		return -1;
+	} else if (ms >= dtlsTransport->mFinMs) {
```
Fixed!

Sean-Der commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,305 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+		return -1;
+	} else if (ms >= dtlsTransport->mFinMs) {
+		return 2;
+	} else if (ms >= dtlsTransport->mIntMs) {
```
Fixed!

Sean-Der commented:
Ok addressed all comments @paullouisageneau 

re: tagging a release. We have a little bit of a chicken and egg problem

* I need to get a version into obs-deps so people can test
* I can't fully know if things are going to work across all platforms, so might need to make more libdatachannel changes

I would like to get a release so I can move ahead with OBS, but I might need to come back with other fixes if I hit something. Would you want to do a beta or rc tag? Does that make things easier on your side

RytoEX commented:
Following up on [this review thread](https://github.com/paullouisageneau/libdatachannel/pull/822#discussion_r1132629334), posting as a standalone comment for visibility:
> Following up on this, I noticed that the finder was initially pulled from the linked repository/PR. While the finder included here has since been modified, if an explicitly licensed file is preferred as the basis of the work here (the finder previously linked to did not have an explicit license), you can find MIT-licensed finders [here](https://github.com/PatTheMav/cmake-finders).


> * I need to get a version into obs-deps so people can test

On the obs-deps side, our WIP PR for this effort is currently just pulling in a commit from your fork instead of from this repo. We will continue to do that until we no longer need to.

Sean-Der commented:
@RytoEX Fixed the Mbed TLS finder. The diff contained no functional changes, but now attributes properly.

paullouisageneau commented:
> I would like to get a release so I can move ahead with OBS, but I might need to come back with other fixes if I hit something. Would you want to do a beta or rc tag? Does that make things easier on your side

@Sean-Der A beta tag sounds fine to me!

paullouisageneau commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,300 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+
+			auto bufMin = std::min(len, size_t(message->size()));
+			std::memcpy(buf, message->data(), bufMin);
+			return len;
```
```suggestion
			return int(len);
```

Sean-Der commented:
@paullouisageneau done! thank you so much for the fast reviews

paullouisageneau commented on src/impl/dtlstransport.cpp:
```diff
@@ -360,7 +360,300 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+
+	int ret;
+	do {
+		std::lock_guard lock(mMutex);
```
The same mutex is missing from `TlsTransport`. It looks like Mbed TLS indeed can't send and receive concurrently on the same session, which is quite a pain.

Sean-Der replied:
Fixed. Updated with locking pattern from DTLS

paullouisageneau commented on src/impl/tlstransport.cpp:
```diff
@@ -296,7 +296,218 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+				if (ret == MBEDTLS_ERR_SSL_WANT_READ || ret == MBEDTLS_ERR_SSL_WANT_WRITE ||
+				    ret == MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS ||
+				    ret == MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS) {
+					continue;
```
You must not call again on `WANT_READ` or `WANT_WRITE` as it would spin when no incoming data is available.

As I understand, you should call again immediately on `X_IN_PROGRESS` and return on `WANT_READ` or `WANT_WRITE` (while scheduling after the timeout for DTLS).

This should be the behavior for each call to `mbedtls_ssl_handshake` and `mbedtls_ssl_read`.

Sean-Der replied:
@paullouisageneau If I remove those all websocket tests fail with. I also get DTLS failures of the same sort

```
2023-03-29 21:09:26.016 ERROR [130773] [rtc::mbedtls::check@106] MbedTLS error: SSL - No data of requested type currently available on underlying transport
2023-03-29 21:09:26.016 ERROR [130773] [rtc::impl::TlsTransport::doRecv@458] TLS recv: MbedTLS error: SSL - No data of requested type currently available on underlying transport
```

paullouisageneau replied:
Hum, `SSL - No data of requested type currently available on underlying transport` is actually `MBEDTLS_ERR_SSL_WANT_READ`, so it would mean `MBEDTLS_ERR_SSL_WANT_READ` ends up in `mbedtls::check()` instead of triggering a return.

Sean-Der commented on src/impl/tlstransport.hpp:
```diff
@@ -65,6 +65,21 @@ class TlsTransport : public Transport, public std::enable_shared_from_this<TlsTr
 	static ssize_t ReadCallback(gnutls_transport_ptr_t ptr, void *data, size_t maxlen);
 	static int TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int ms);
+#elif USE_MBEDTLS
+	std::mutex mSendMutex;
```
Fixed. Updated with locking pattern from DTLS

paullouisageneau commented:
I think the last issue is with `mbedtls_ssl_handshake` and `mbedtls_ssl_read` return value handling [here](https://github.com/paullouisageneau/libdatachannel/pull/822#discussion_r1152465327), otherwise it looks good to me!

After fixing the issue, I'll merge it and create a tag.

Sean-Der commented:
Ok fixed! @paullouisageneau fingers crossed I think I got it all right :D 

Thanks for all your patience on your PR. Getting WHIP into OBS is going to be so great for people I can't wait.

If someone can broadcast to their friends/co-workers/x$ without setting up a expensive server it is going to be massive for people. Really excited.

paullouisageneau commented:
> Ok fixed! @paullouisageneau fingers crossed I think I got it all right :D

@Sean-Der Looks good, could you just fix the DTLS handshake and read too before merging?

Thank you for your work and patience! WHIP in OBS is indeed very exciting, but I know Mbed TLS support will also unlock new possibilities for a lot of people.

Sean-Der commented:
@paullouisageneau done! I think I got it this time :D

paullouisageneau reviewed (approved):
It looks good, thank you!

paullouisageneau commented:
@Sean-Der Tagged as ` 0.19.0-alpha.1`

pkviet commented:
Thanks from obs devs. @paullouisageneau
